src/AggregateVersioning/AggregateVersionListener.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Gedmo\AggregateVersioning;
+
+use Doctrine\Common\Annotations\AnnotationReader;
+use Doctrine\ORM\Event\OnFlushEventArgs;
+use Gedmo\Mapping\Annotation\AggregateVersioning;
+use Gedmo\Mapping\MappedEventSubscriber;
+use LogicException;
+use ReflectionClass;
+
+/**
+ * The Aggregate version listener handles the updated Aggregate Root and Aggregate Entity
+ * and update Aggregate version
+ *
+ * @author Maksim Vorozhtsov <myks1992@mail.ru>
+ * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
+ */
+final class AggregateVersionListener extends MappedEventSubscriber
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getSubscribedEvents(): array
+    {
+        return [
+            'onFlush',
+        ];
+    }
+
+    public function onFlush(OnFlushEventArgs $args): void
+    {
+        $em = $args->getEntityManager();
+        $uow = $em->getUnitOfWork();
+
+        $all = array_merge(
+            $uow->getScheduledEntityInsertions(),
+            $uow->getScheduledEntityUpdates(),
+            $uow->getScheduledEntityDeletions()
+        );
+
+        $entities = [];
+        foreach ($all as $entity) {
+            if ($entity instanceof AggregateEntity || $entity instanceof AggregateRoot) {
+                $entities[] = $entity;
+            }
+        }
+
+        if (empty($entities)) {
+            return;
+        }
+
+        $aggregateRoot = $this->getAggregateRoot($entities);
+        $aggregateRoot->updateAggregateVersion();
+
+        $meta = $em->getClassMetadata(get_class($aggregateRoot));
+        $uow->recomputeSingleEntityChangeSet($meta, $aggregateRoot);
+    }
+
+    private function getAggregateRoot(array $entities): AggregateRoot
+    {
+        foreach ($entities as $entity) {
+            if ($entity instanceof AggregateRoot) {
+                return $entity;
+            }
+        }
+
+        $entity = end($entities);
```

Comment:
Even if we currently know that the only method calling this one will not pass an empty array as argument 1, we should take care of that case.